spring/boot-autoconfigure/src/main/java/com/linecorp/armeria/spring/AbstractServiceRegistrationBean.java

```diff
@@ -0,0 +1,133 @@
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.server.Service;
+
+/**
+ * An abstract bean with information for registering an annotated service object. It enables micrometer
+ * monitoring of the service automatically.
+ */
+class AbstractServiceRegistrationBean<T, U> {
+    /**
+     * The annotated service object to register.
+     */
+    @NotNull
+    private T service;
+
+    /**
+     * A service name to use in monitoring.
+     */
+    @NotNull
+    private String serviceName;
+
+    /**
+     * The decorators of the annotated service object.
+     */
+    @NotNull
+    private List<Function<Service<HttpRequest, HttpResponse>,
+            ? extends Service<HttpRequest, HttpResponse>>> decorators = new ArrayList<>();
+
+    /**
+     * Returns the annotated service object registered to this bean.
+     */
+    @NotNull
+    public T getService() {
+        return service;
+    }
+
+    /**
+     * Registers an annotated service object.
+     */
+    public U setService(@NotNull T service) {
+        this.service = service;
+        return self();
+    }
+
+    /**
+     * Returns this service name to use in monitoring.
+     */
+    @NotNull
+    public String getServiceName() {
+        return serviceName;
+    }
+
+    /**
+     * Sets service name to use in monitoring.
+     */
+    public U setServiceName(
+            @NotNull String serviceName) {
+        this.serviceName = serviceName;
+        return self();
+    }
+
+    /**
+     * Returns the decorators of the annotated service object.
+     */
+    @NotNull
+    public List<Function<Service<HttpRequest, HttpResponse>,
+            ? extends Service<HttpRequest, HttpResponse>>> getDecorators() {
+        return decorators;
+    }
+
+    /**
+     * Sets the decorator of the annotated service object. {@code decorator} are applied to {@code service} in
+     * order.
+     * @deprecated Use {@link #setDecorators(Function[])} or {@link #setDecorators(List)} instead.
+     */
+    @Deprecated
+    public U setDecorator(
+            Function<Service<HttpRequest, HttpResponse>,
+                    ? extends Service<HttpRequest, HttpResponse>>... decorators) {
+        return setDecorators(decorators);
+    }
+
+    /**
+     * Sets the decorator of the annotated service object. {@code decorators} are applied to {@code service} in
+     * order.
+     */
+    public U setDecorators(
+            Function<Service<HttpRequest, HttpResponse>,
+                    ? extends Service<HttpRequest, HttpResponse>>... decorators) {
+        return setDecorators(ImmutableList.copyOf(decorators));
+    }
+
+    /**
+     * Sets the decorators of the annotated service object. {@code decorators} are applied to {@code service} in
+     * order.
+     */
+    public U setDecorators(
+            List<Function<Service<HttpRequest, HttpResponse>,
+                    ? extends Service<HttpRequest, HttpResponse>>> decorators) {
+        this.decorators = decorators;
```

Comment:
I don't know if we need `requireNonNull()` here. If we don't, just ignore this comment. :)